src/components/implementation/tests/micro_booter/mb_tests.c

```diff
@@ -499,267 +453,299 @@ test_budgets(void)
-	PRINTC(" - SUCCESS.\n");
 }
 
 static void
-test_preemption(void)
+test_async_endpoints_perf(void)
 {
-	PRINTC("Testing Preemption\n");
-	exec_cluster_alloc(&pat[cos_cpuid()].i, sender_thd, &pat[cos_cpuid()], BOOT_CAPTBL_SELF_INITRCV_CPU_BASE);
-	exec_cluster_alloc(&pat[cos_cpuid()].w, receiver_thd, &pat[cos_cpuid()], BOOT_CAPTBL_SELF_INITRCV_CPU_BASE);
+	thdcap_t  tcp, tcc;
+	tcap_t    tccp, tccc;
+	arcvcap_t rcp, rcc;
 
-	/*
-	 * test case 1: Sender = H, Receiver: L
-	 * - scheduler here is initthd (this thd)
-	 * expected result:
-	 * - cos_asnd from sender should add receiver as wakeup thread.
-	 * - cos_rcv from sender should activate receiver thread.
-	 */
-	PRINTC(" Test - Sender = H, Receiver = L");
-	test_preemption_case(&pat[cos_cpuid()], TEST_PRIO_HIGH, TEST_PRIO_LOW, 0, 1);
+	/* parent rcv capabilities */
+	tcp = cos_thd_alloc(&booter_info, booter_info.comp_cap, async_thd_parent_perf,
+	                    (void *)BOOT_CAPTBL_SELF_INITTHD_CPU_BASE);
+	if(EXPECT_LL_LT(1, tcp, "Test Async Endpoints"))
+		return;
+	tccp = cos_tcap_alloc(&booter_info);
+	if(EXPECT_LL_LT(1, tccp, "Test Async Endpoints"))
+		return;
+	rcp = cos_arcv_alloc(&booter_info, tcp, tccp, booter_info.comp_cap, BOOT_CAPTBL_SELF_INITRCV_CPU_BASE);
+	if(EXPECT_LL_LT(1, rcp, "Test Async Endpoints"))
+		return;
+	if(EXPECT_LL_NEQ(0,cos_tcap_transfer(rcp, BOOT_CAPTBL_SELF_INITTCAP_CPU_BASE, TCAP_RES_INF, TCAP_PRIO_MAX + 1), "Test Async Endpoints")) {
+		return;
+	}
 
-	/*
-	 * test case 2: Sender = L, Receiver: H
-	 * - scheduler here is initthd (this thd)
-	 * expected result:
-	 * - cos_asnd from sender should trigger receiver activation and add sender as wakeup thread.
-	 * - cos_rcv from receiver should activate sender thread.
-	 */
-	PRINTC(" Test - Sender = L, Receiver = H");
-	test_preemption_case(&pat[cos_cpuid()], TEST_PRIO_LOW, TEST_PRIO_HIGH, 1, 0);
+	/* child rcv capabilities */
+	tcc = cos_thd_alloc(&booter_info, booter_info.comp_cap, async_thd_fn_perf, (void *)tcp);
+	if(EXPECT_LL_LT(1, tcc, "Test Async Endpoints"))
+		return;
+	tccc = cos_tcap_alloc(&booter_info);
+	if(EXPECT_LL_LT(1, tccc, "Test Async Endpoints"))
+		return;
+	rcc = cos_arcv_alloc(&booter_info, tcc, tccc, booter_info.comp_cap, rcp);
+	if(EXPECT_LL_LT(1, rcc, "Test Async Endpoints"))
+		return;
+	if(EXPECT_LL_NEQ(0,cos_tcap_transfer(rcc, BOOT_CAPTBL_SELF_INITTCAP_CPU_BASE, TCAP_RES_INF, TCAP_PRIO_MAX), "Test Async Endpoints"))
+		 return;
+
+	/* make the snd channel to the child */
+	scp_global[cos_cpuid()] = cos_asnd_alloc(&booter_info, rcc, booter_info.captbl_cap);
+	if(EXPECT_LL_EQ(0, scp_global[cos_cpuid()], "Test Async Endpoints"))
+		return;
 
-	PRINTC("Done.\n");
+	rcc_global[cos_cpuid()] = rcc;
+	rcp_global[cos_cpuid()] = rcp;
+
+	async_test_flag_[cos_cpuid()] = 1;
+	while (async_test_flag_[cos_cpuid()])
+        cos_thd_switch(tcp);
 }
 
-long long midinv_cycles[NUM_CPU] = { 0LL };
+long long midinv_cycle[NUM_CPU] = { 0LL };
```

Comment:
We use "cycles" everywhere. ;-)

Reply:
Long story short midinv_cycles is declared inside a component. `Redeclaration` is thrown by the compiler